kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_constraints.h

```diff
@@ -0,0 +1,1007 @@
+
+        const int nconditions = static_cast<int>(rModelPart.Conditions().size());
+#pragma omp parallel for firstprivate(nconditions, ids)
+        for (int iii = 0; iii < nconditions; iii++)
+        {
+            typename ConditionsArrayType::iterator i_condition = rModelPart.Conditions().begin() + iii;
+            pScheme->Condition_EquationId(*(i_condition.base()), ids, rModelPart.GetProcessInfo());
+            ApplyConstraints<Condition>(rModelPart, *i_condition, ids, rModelPart.GetProcessInfo());
+            for (std::size_t i = 0; i < ids.size(); i++)
+            {
+#ifdef _OPENMP
+                omp_set_lock(&BaseType::mlock_array[ids[i]]);
+#endif
+                auto &row_indices = indices[ids[i]];
+                row_indices.insert(ids.begin(), ids.end());
+#ifdef _OPENMP
+                omp_unset_lock(&BaseType::mlock_array[ids[i]]);
+#endif
+            }
+        }
+
+        //count the row sizes
+        unsigned int nnz = 0;
+        for (unsigned int i = 0; i < indices.size(); i++)
+            nnz += indices[i].size();
+
+        A = boost::numeric::ublas::compressed_matrix<double>(indices.size(), indices.size(), nnz);
+
+        double *Avalues = A.value_data().begin();
+        std::size_t *Arow_indices = A.index1_data().begin();
+        std::size_t *Acol_indices = A.index2_data().begin();
+
+        //filling the index1 vector - DO NOT MAKE PARALLEL THE FOLLOWING LOOP!
+        Arow_indices[0] = 0;
+        for (int i = 0; i < static_cast<int>(A.size1()); i++)
+            Arow_indices[i + 1] = Arow_indices[i] + indices[i].size();
+
+#pragma omp parallel for
+        for (int i = 0; i < static_cast<int>(A.size1()); i++)
+        {
+            const unsigned int row_begin = Arow_indices[i];
+            const unsigned int row_end = Arow_indices[i + 1];
+            unsigned int k = row_begin;
+            for (auto it = indices[i].begin(); it != indices[i].end(); it++)
+            {
+                Acol_indices[k] = *it;
+                Avalues[k] = 0.0;
+                k++;
+            }
+
+            indices[i].clear(); //deallocating the memory
+
+            std::sort(&Acol_indices[row_begin], &Acol_indices[row_end]);
+        }
+
+        A.set_filled(indices.size() + 1, nnz);
+
+        Timer::Stop("MatrixStructure");
+    }
+
+    ///@}
+    ///@name Protected  Access
+    ///@{
+
+    ///@}
+    ///@name Protected Inquiry
+    ///@{
+
+    ///@}
+    ///@name Protected LifeCycle
+    ///@{
+
+    ///@}
+
+  private:
+    ///@name Static Member Variables
+    ///@{
+
+    ///@}
+    ///@name Member Variables
+    ///@{
+
+    // This is the set of condenced global constraints.
+    GlobalMasterSlaveRelationContainerType mGlobalMasterSlaveRelations; //This can be changed to more efficient implementation later on.
+
+    ///@}
+    ///@name Private Operators
+    ///@{
+
+    ///@}
+    ///@name Private Operations
+    ///@{
+
+    void FormulateGlobalMasterSlaveRelations(ModelPart &rModelPart)
```

Comment:
Just one thing, can you add some commentaries to these methods? (The name is self explanatory, but I thing that it can be a little bit expanded)